components/skin/ui/src/main/resources/PhenoTips/Skin.xml

```diff
@@ -507,39 +507,144 @@ document.observe('xwiki:suggest:containerCreated', function(event) {
+        // if it did.
+        window.addEventListener('beforeunload', function(event) {
+          if (_this.isFormDirty()) {
+            var message = "$!escapetool.javascript($services.localization.render('phenotips.navigation.beforeunload.warning'))";
+            (event || window.event).returnValue = message;
+            return message;
+          }
+          return undefined;
+        });
+
+        // Set up the smart autosave
+        if (this.config.autosave.enabled) {
+          this.timer = new PeriodicalExecuter(this.saveIfFormDirty.bind(this), this.config.autosave.frequency);
+        }
+      }
+    },
+    isFormDirty : function() {
+      return this.form ? (this.formData != this.form.serialize()) : false;
+    },
+    saveForm : function(successCallback, failureCallback) {
+      document.fire("xwiki:actions:save", {"continue": true, "form" : this.form});
+      // If there's a callback for there's no unsaved data:
+      if (typeof successCallback == 'function') {
+        // Call the callback function
+        document.observe("xwiki:document:saved", function onSaved() {
+          document.stopObserving("xwiki:document:saved", onSaved);
+          successCallback();
+        });
+        // At the same time, if there's NO callback for when the saving failed, ask user if it's OK to proceed anyway with the successCallback
+        if (typeof failureCallback != 'function') {
+          document.observe("xwiki:document:saveFailed", function onSaveFailed() {
+            document.stopObserving("xwiki:document:saveFailed", onSaveFailed);
+            (new XWiki.widgets.ConfirmationBox({
+              onYes: successCallback,
+              onNo: function(){}
+            },
+            {
+              confirmationText: "$!escapetool.javascript($services.localization.render('phenotips.save.failed.confirmProceed'))", // "Failed to save the latest modifications. Proceed anyway?"
+              showCancelButton: false
+             })).showDialog();
+          });
         }
-        return undefined;
+      }
+      // However, if there is a callback for when the saving failed, call that one at the saveFailed event
+      (typeof failureCallback == 'function') &amp;&amp; document.observe("xwiki:document:saveFailed", function onSaveFailed() {
+        document.stopObserving("xwiki:document:saveFailed", onSaveFailed);
+        failureCallback();
       });
+    },
+    saveIfFormDirty : function(successCallback, failureCallback) {
+      if (this.isFormDirty()) {
+        this.saveForm(successCallback, failureCallback);
+      } else {
+        // If there are no unsaved changes, just go straight to calling the successCallback function if one is provided.
+        (typeof successCallback == 'function') &amp;&amp; successCallback();
+      }
     }
+  });
+
+  // Static function to get the form state
+  widgets.FormUtils.getFormState = function() {
+    var form = $('inline');
+    return form &amp;&amp; form._formState;
```

Comment:
This violates the documentation written above:

> if there is no $('inline') form on the page, getFormState will still return an object

Which means that users of this method will have to check first if the state is non-null. Please make this conform to the written documentation.